### PR TITLE
Improves the API for monitoring contacts in RigidDynamicBody

### PR DESCRIPTION
**Pull request by fabriceci**
- Rename "contacts_reported" to `max_contacts_reported` as the name is confusing.
- Complete the documentation to explain the flow
- Add a `get_contact_count()` method to get the number of contacts

### CODE REVIEW

File: doc/classes/RigidDynamicBody2D.xml [hunk 1]
```diff
@@ -187,6 +191,10 @@
 		<member name="mass" type="float" setter="set_mass" getter="get_mass" default="1.0">
 			The body's mass.
 		</member>
+		<member name="max_contacts_reported" type="int" setter="set_max_contacts_reported" getter="get_max_contacts_reported" default="0">
+			The maximum number of contacts that will be recorded. Requires a value greater than 0 and [member contact_monitor] to be set to [code]true[/code] to starting to register contacts. Use [method get_contact_count] to retrieve the count or [method get_contact_count] to retrieve bodies that have collided with.
```
Comment (rburing) [inline doc/classes/RigidDynamicBody2D.xml:195]:
```suggestion
			The maximum number of contacts that will be recorded. Requires a value greater than 0 and [member contact_monitor] to be set to [code]true[/code] to start to register contacts. Use [method get_contact_count] to retrieve the count or [method get_colliding_bodies] to retrieve bodies that have been collided with.
```

File: doc/classes/RigidDynamicBody2D.xml [hunk 2]
```diff
@@ -100,10 +100,17 @@
...
 			<description>
-				Returns a list of the bodies colliding with this one. Requires [member contact_monitor] to be set to [code]true[/code] and [member contacts_reported] to be set high enough to detect all the collisions.
+				Returns a list of the bodies colliding with this one. Requires [member contact_monitor] to be set to [code]true[/code] and [member max_contacts_reported] to be set high enough to detect all the collisions.
 				[b]Note:[/b] The result of this test is not immediate after moving objects. For performance, list of collisions is updated once per frame and before the physics step. Consider using signals instead.
 			</description>
 		</method>
+		<method name="get_contact_count" qualifiers="const">
+			<return type="int" />
+			<description>
+				Returns the number of contacts this body has with other bodies.
```
Comment (rburing) [inline doc/classes/RigidDynamicBody2D.xml:110]: A note about the difference between contacts and collisions could go here (as well).

File: doc/classes/RigidDynamicBody3D.xml [hunk 3]
```diff
@@ -193,6 +197,10 @@
 		<member name="mass" type="float" setter="set_mass" getter="get_mass" default="1.0">
 			The body's mass.
 		</member>
+		<member name="max_contacts_reported" type="int" setter="set_max_contacts_reported" getter="get_max_contacts_reported" default="0">
+			The maximum number of contacts that will be recorded. Requires a value greater than 0 and [member contact_monitor] to be set to [code]true[/code] to starting to register contacts. Use [method get_contact_count] to retrieve the count or [method get_contact_count] to retrieve bodies that have collided with.
```
Comment (rburing) [inline doc/classes/RigidDynamicBody3D.xml:201]:
```suggestion
			The maximum number of contacts that will be recorded. Requires a value greater than 0 and [member contact_monitor] to be set to [code]true[/code] to start to register contacts. Use [method get_contact_count] to retrieve the count or [method get_colliding_bodies] to retrieve bodies that have been collided with.
```

File: doc/classes/RigidDynamicBody3D.xml [hunk 4]
```diff
@@ -100,10 +100,17 @@
...
 			<description>
-				Returns a list of the bodies colliding with this one. Requires [member contact_monitor] to be set to [code]true[/code] and [member contacts_reported] to be set high enough to detect all the collisions.
+				Returns a list of the bodies colliding with this one. Requires [member contact_monitor] to be set to [code]true[/code] and [member max_contacts_reported] to be set high enough to detect all the collisions.
 				[b]Note:[/b] The result of this test is not immediate after moving objects. For performance, list of collisions is updated once per frame and before the physics step. Consider using signals instead.
 			</description>
 		</method>
+		<method name="get_contact_count" qualifiers="const">
+			<return type="int" />
+			<description>
+				Returns the number of contacts this body has with other bodies.
```
Comment (rburing) [inline doc/classes/RigidDynamicBody3D.xml:110]: A note about the difference between contacts and collisions could go here (as well).

**Review by rburing — APPROVED** [review]
Makes sense to me!

File: doc/classes/RigidDynamicBody2D.xml [hunk 5]
```diff
@@ -100,10 +100,17 @@
...
-				Returns a list of the bodies colliding with this one. Requires [member contact_monitor] to be set to [code]true[/code] and [member contacts_reported] to be set high enough to detect all the collisions.
+				Returns a list of the bodies colliding with this one. Requires [member contact_monitor] to be set to [code]true[/code] and [member max_contacts_reported] to be set high enough to detect all the collisions.
 				[b]Note:[/b] The result of this test is not immediate after moving objects. For performance, list of collisions is updated once per frame and before the physics step. Consider using signals instead.
 			</description>
 		</method>
+		<method name="get_contact_count" qualifiers="const">
+			<return type="int" />
+			<description>
+				Returns the number of contacts this body has with other bodies. By default, this returns 0 unless bodies are configured to monitor contacts (see [member contact_monitor]).
+				[b]Note:[/b] to retrieve the colliding bodies, use [method get_colliding_bodies]
```
Comment (mhilbrunner) [inline doc/classes/RigidDynamicBody2D.xml:111]:
```suggestion
				[b]Note:[/b] to retrieve the colliding bodies, use [method get_colliding_bodies].
```

**Review by mhilbrunner — APPROVED** [review]
Literally the smallest possible nitpick, otherwise LGTM.

File: doc/classes/RigidDynamicBody2D.xml (hunk 5 above)
Comment (akien-mga) [reply doc/classes/RigidDynamicBody2D.xml:111]: Second nitpick: the sentence after "Note:" should start with an capitalized word.

File: doc/classes/RigidDynamicBody3D.xml [hunk 6]
```diff
@@ -100,10 +100,17 @@
...
-				Returns a list of the bodies colliding with this one. Requires [member contact_monitor] to be set to [code]true[/code] and [member contacts_reported] to be set high enough to detect all the collisions.
+				Returns a list of the bodies colliding with this one. Requires [member contact_monitor] to be set to [code]true[/code] and [member max_contacts_reported] to be set high enough to detect all the collisions.
 				[b]Note:[/b] The result of this test is not immediate after moving objects. For performance, list of collisions is updated once per frame and before the physics step. Consider using signals instead.
 			</description>
 		</method>
+		<method name="get_contact_count" qualifiers="const">
+			<return type="int" />
+			<description>
+				Returns the number of contacts this body has with other bodies. By default, this returns 0 unless bodies are configured to monitor contacts (see [member contact_monitor]).
+				[b]Note:[/b] to retrieve the colliding bodies, use [method get_colliding_bodies]
```
Comment (akien-mga) [inline doc/classes/RigidDynamicBody3D.xml:111]: Same comments as for 2D.

**Comment by akien-mga** [comment]
Thanks!

**Comment by dmaz** [comment]
As far as I can see this rename didn't auto update so it looks to me like this was a breaking change.  Any node where you had this set the setting was just removed

**Pull request merged**